models/unit/unit.go

```diff
@@ -154,9 +154,13 @@ func LoadUnitConfig() {
 	DisabledRepoUnits = FindUnitTypes(setting.Repository.DisabledRepoUnits...)
 	// Check that must units are not disabled
 	for i, disabledU := range DisabledRepoUnits {
-		if !disabledU.CanDisable() {
+		if !disabledU.CanDisable() || disabledU.IsInvalid() {
 			log.Warn("Not allowed to global disable unit %s", disabledU.String())
-			DisabledRepoUnits = append(DisabledRepoUnits[:i], DisabledRepoUnits[i+1:]...)
+			if i == 0 {
+				DisabledRepoUnits = DisabledRepoUnits[i:]
+			} else {
+				DisabledRepoUnits = append(DisabledRepoUnits[:i], DisabledRepoUnits[i+1:]...)
+			}
```

Comment:
`DisabledRepoUnits = append(DisabledRepoUnits[:i], DisabledRepoUnits[i+1:]...)` really seems strange and fragile.


I think you need something like:

```
y := x[:0]
for _, n := range x {
    if... {
        y = append(y, n)
    }
}
```

Maybe:

```
disabledRepoUnits := FindUnitTypes(setting.Repository.DisabledRepoUnits...)
DisabledRepoUnits = disabledRepoUnits
for _, u := range disabledRepoUnits {
      if ... {
            DisabledRepoUnits = append(DisabledRepoUnits, u)
      }
}
```

Reply:
This code can be found in golang's wiki, but it can not be used in `for x range x`
https://github.com/golang/go/wiki/SliceTricks#delete
I found a discussion about this in stackoverflow
https://stackoverflow.com/questions/20545743/how-to-remove-items-from-a-slice-while-ranging-over-it
![image](https://user-images.githubusercontent.com/18380374/220297116-ec2fbe86-be67-40a3-aefc-ae6e3dd020eb.png)

@wxiaoguang 
I will try your code :) thanks for your suggestion.

Reply:
If you only remove one, that's the best app. But if you are removing many, I think the new variable is better, and it's still effenient, no memory allocation.

Reply:
OK, I get my computer now. Here is a demo about why you shouldn't use the old code

```go
package main

func main() {
	DisabledRepoUnits := []int{1, 2, 3, 4}
	for i, disabledU := range DisabledRepoUnits {
		if disabledU%2 == 0 {
			if i == 0 {
				DisabledRepoUnits = DisabledRepoUnits[i:]
			} else {
				DisabledRepoUnits = append(DisabledRepoUnits[:i], DisabledRepoUnits[i+1:]...)
			}
		}
	}
	println(DisabledRepoUnits)
}
```

It just panics:


```
panic: runtime error: slice bounds out of range [4:2]

goroutine 1 [running]:
main.main()
	/tmp/sandbox15630008/prog.go:12 +0x1ec
```